File: data-structures/go/src/dynamicarray/dynamic_array.go
package dynamicarray

import "errors"

var ErrOutOfRange = errors.New("DynamicArray: index out of range")
var ErrEmpty = errors.New("DynamicArray: array is empty")

type DynamicArray[T any] struct {
	data     []T
	size     int
	capacity int
}

func New[T any]() *DynamicArray[T] {
	return &DynamicArray[T]{}
}

func NewWithSize[T any](size int) *DynamicArray[T] {
	if size <= 0 {
		return &DynamicArray[T]{}
	}
	return &DynamicArray[T]{
		data:     make([]T, size),
		size:     size,
		capacity: size,
	}
}

func (a *DynamicArray[T]) At(index int) (T, error) {
	var zero T
	if index < 0 || index >= a.size {
		return zero, ErrOutOfRange
	}
	return a.data[index], nil
}

func (a *DynamicArray[T]) SetAt(index int, value T) error {
	if index < 0 || index >= a.size {
		return ErrOutOfRange
	}
	a.data[index] = value
	return nil
}

func (a *DynamicArray[T]) Get(index int) T {
	return a.data[index]
}

func (a *DynamicArray[T]) Set(index int, value T) {
	a.data[index] = value
}

func (a *DynamicArray[T]) Front() (T, error) {
	var zero T
	if a.size == 0 {
		return zero, ErrEmpty
	}
	return a.data[0], nil
}

func (a *DynamicArray[T]) Back() (T, error) {
	var zero T
	if a.size == 0 {
		return zero, ErrEmpty
	}
	return a.data[a.size-1], nil
}

func (a *DynamicArray[T]) Data() []T {
	if a.size == 0 {
		return nil
	}
	return a.data[:a.size]
}

func (a *DynamicArray[T]) Size() int {
	return a.size
}

func (a *DynamicArray[T]) Capacity() int {
	return a.capacity
}

func (a *DynamicArray[T]) IsEmpty() bool {
	return a.size == 0
}

func (a *DynamicArray[T]) Reserve(newCap int) {
	if newCap <= a.capacity {
		return
	}
	newData := make([]T, newCap)
	copy(newData, a.data[:a.size])
	a.data = newData
	a.capacity = newCap
}

func (a *DynamicArray[T]) PushBack(value T) {
	if a.size == a.capacity {
		newCap := 1
		if a.capacity > 0 {
			newCap = a.capacity * 2
		}
		a.Reserve(newCap)
	}
	a.data[a.size] = value
	a.size++
}

func (a *DynamicArray[T]) PopBack() error {
	if a.size == 0 {
		return ErrEmpty
	}
	a.size--
	var zero T
	a.data[a.size] = zero
	return nil
}

func (a *DynamicArray[T]) Clear() {
	var zero T
	for i := 0; i < a.size; i++ {
		a.data[i] = zero
	}
	a.size = 0
}

func (a *DynamicArray[T]) Clone() *DynamicArray[T] {
	clone := &DynamicArray[T]{
		data:     make([]T, a.capacity),
		size:     a.size,
		capacity: a.capacity,
	}
	copy(clone.data, a.data[:a.size])
	return clone
}

File: data-structures/python/src/dynamic_array.py
class DynamicArray:
    def __init__(self, size=0):
        if not isinstance(size, int) or size < 0:
            raise ValueError("size must be a non-negative integer")
        self._size = size
        self._capacity = size
        self._data = [None] * size if size > 0 else []

    def at(self, index):
        if index < 0 or index >= self._size:
            raise IndexError("DynamicArray.at: index out of range")
        return self._data[index]

    def set_at(self, index, value):
        if index < 0 or index >= self._size:
            raise IndexError("DynamicArray.set_at: index out of range")
        self._data[index] = value

    def __getitem__(self, index):
        return self._data[index]

    def __setitem__(self, index, value):
        self._data[index] = value

    def front(self):
        if self._size == 0:
            raise IndexError("DynamicArray.front: array is empty")
        return self._data[0]

    def back(self):
        if self._size == 0:
            raise IndexError("DynamicArray.back: array is empty")
        return self._data[self._size - 1]

    def data(self):
        if self._size == 0:
            return None
        return self._data[:self._size]

    def size(self):
        return self._size

    def capacity(self):
        return self._capacity

    def empty(self):
        return self._size == 0

    def reserve(self, new_cap):
        if new_cap <= self._capacity:
            return
        new_data = [None] * new_cap
        for i in range(self._size):
            new_data[i] = self._data[i]
        self._data = new_data
        self._capacity = new_cap

    def push_back(self, value):
        if self._size == self._capacity:
            new_cap = 1 if self._capacity == 0 else self._capacity * 2
            self.reserve(new_cap)
        self._data[self._size] = value
        self._size += 1

    def pop_back(self):
        if self._size == 0:
            raise IndexError("DynamicArray.pop_back: array is empty")
        self._size -= 1

    def clear(self):
        self._size = 0

    def copy(self):
        clone = DynamicArray()
        clone._size = self._size
        clone._capacity = self._capacity
        clone._data = [None] * self._capacity
        for i in range(self._size):
            clone._data[i] = self._data[i]
        return clone

    def __iter__(self):
        for i in range(self._size):
            yield self._data[i]

    def __len__(self):
        return self._size
